go-controller/pkg/libovsdbops/router.go

```diff
@@ -44,6 +44,428 @@ func FindLogicalRoutersWithPredicate(nbClient libovsdbclient.Client, p logicalRo
+			OnModelMutations: []interface{}{&router.Policies},
+			ErrNotFound:      true,
+			BulkOp:           false,
+		},
+	}
+
+	m := NewModelClient(nbClient)
+	_, err := m.CreateOrUpdate(opModels...)
+	return err
+}
+
+// DeleteLogicalRouterPoliciesWithPredicate looks up logical router policies
+// from the cache based on a given predicate, deletes them and removes them from
+// the provided logical router
+func DeleteLogicalRouterPoliciesWithPredicate(nbClient libovsdbclient.Client, routerName string, p logicalRouterPolicyPredicate) error {
+	router := &nbdb.LogicalRouter{
+		Name: routerName,
+	}
+
+	deleted := []*nbdb.LogicalRouterPolicy{}
+	opModels := []OperationModel{
+		{
+			ModelPredicate: p,
+			ExistingResult: &deleted,
+			DoAfter:        func() { router.Policies = ExtractUUIDsFromModels(&deleted) },
+			ErrNotFound:    false,
+			BulkOp:         true,
+		},
+		{
+			Model:            router,
+			ModelPredicate:   func(lr *nbdb.LogicalRouter) bool { return lr.Name == routerName },
+			OnModelMutations: []interface{}{&router.Policies},
+			ErrNotFound:      true,
+			BulkOp:           false,
+		},
+	}
+
+	m := NewModelClient(nbClient)
+	return m.Delete(opModels...)
+}
+
+// CreateOrAddNextHopsToLogicalRouterPolicyWithPredicate looks up a logical
+// router policy from the cache based on a given predicate. If it doesn't find
+// any, it creates the provided logical router policy. If it does, adds any
+// missing Nexthops to the existing logical router policy. The logical router
+// policy is added to the provided logical router.
+func CreateOrAddNextHopsToLogicalRouterPolicyWithPredicate(nbClient libovsdbclient.Client, routerName string, lrp *nbdb.LogicalRouterPolicy, p logicalRouterPolicyPredicate) error {
+	router := &nbdb.LogicalRouter{
+		Name: routerName,
+	}
+
+	opModels := []OperationModel{
+		{
+			Model:            lrp,
+			ModelPredicate:   p,
+			OnModelMutations: []interface{}{&lrp.Nexthops},
+			DoAfter:          func() { router.Policies = []string{lrp.UUID} },
+			ErrNotFound:      false,
+			BulkOp:           false,
+		},
+		{
+			Model:            router,
+			ModelPredicate:   func(item *nbdb.LogicalRouter) bool { return item.Name == router.Name },
+			OnModelMutations: []interface{}{&router.Policies},
+			ErrNotFound:      true,
+			BulkOp:           false,
+		},
+	}
+
+	m := NewModelClient(nbClient)
+	_, err := m.CreateOrUpdate(opModels...)
+	return err
+}
+
+func deleteNextHopsFromLogicalRouterPolicyOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, routerName string, lrps []*nbdb.LogicalRouterPolicy, nextHops ...string) ([]libovsdb.Operation, error) {
+	nextHopSet := sets.NewString(nextHops...)
+	opModels := []OperationModel{}
+	router := &nbdb.LogicalRouter{
+		Name:     routerName,
+		Policies: []string{},
+	}
+
+	for i := range lrps {
+		lrp := lrps[i]
+		if len(lrp.Nexthops) == len(nextHops) && nextHopSet.HasAll(nextHops...) {
```

Comment:
Should the second condition be  `nextHopSet.HasAll(lrp.Nexthops...)`?
Right now the second part is always true since `nextHopSet := sets.NewString(nextHops...)`